Remove uses of `typeof` and the `?:` operator

romoore opened the pull request:
Make the code base more portable by replacing some GNU extension code
with C99-compatible versions.

* Replace shortcut `?:` operator with explicit middle (*true*) terms
* Remove pointer inequality checks in conditionals
* (Cleanup) Change to `#include` for `<tgmath.h>` based on style guide

Closes #846

ianegordon commented on components/FlexibleHeader/src/MDCFlexibleHeaderView.m:
```diff
@@ -662,24 +663,29 @@ - (void)fhv_accumulatorDidChange {
       // When over non-infinite content we also want to hide the shadow when we're anchored to the
       // top of our content.
-      shadowIntensity = MAX(0, MIN(1, MIN(accumulator, frameBottomEdge) / kShadowScaleLength));
+      shadowIntensity = MDCFMax(0,
```
If you add a linebreak after "shadowIntensity =" will everything fit on the next line?

romoore replied:
Yep, will change.

ianegordon commented on components/PageControl/src/MDCPageControl.m:
```diff
@@ -56,6 +58,16 @@ static inline CGFloat normalizeValue(CGFloat value, CGFloat minRange, CGFloat ma
+/**
+ Snap @c value to the range <code>[min, max]</code>.
+ */
+static inline NSInteger snapToRange(NSInteger min, NSInteger max, NSInteger value) {
```
Functions are typically capitalized.
For minimal surprise, consider naming Clamp(value, min, max) to match other implementations.

romoore replied:
Done.

ianegordon commented on components/TextFields/src/MDCTextInputControllerFullWidth.m:
```diff
@@ -849,7 +851,7 @@ - (void)setErrorText:(NSString *)errorText
     NSString *valueString = @"";
 
-    if (self.textInput.text > 0) {
+    if (self.textInput.text) {
```
Was this originally intended to be:
if (self.textInput.text.length > 0)

willlarche replied:
yesy

ianegordon reviewed (changes requested):
MIN and MAX are defined in NSObjCRuntime.h which is in Foundation.
I'm not sure we should be re-implementing it.
Let's discuss offline.

ianegordon commented:
Discussed offline.
TODO: Investigate if it is simple to enable the extension warning on our code without tripping up on the system code.
Move back to MIN / MAX.

Commit:
Remove uses of `typeof` and the `?:` operator

Make the code base more portable by replacing some GNU extension code
with C99-compatible versions.

* Replace shortcut ?: operator with explicit middle (true) terms
* Remove pointer inequality checks in conditionals
* (Cleanup) Change to #include for <tgmath.h> based on style guide

Closes #846

ianegordon commented on components/TextFields/src/MDCTextField.m:
```diff
@@ -56,8 +56,9 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
-                     ?: [[MDCTextInputCommonFundament alloc] initWithTextInput:self];
+    MDCTextInputCommonFundament *fundament = [aDecoder decodeObjectForKey:MDCTextFieldFundamentKey];
+    _fundament = fundament
+                     ? fundament : [[MDCTextInputCommonFundament alloc] initWithTextInput:self];
```
Consider a linebreak after the "lvar =" to keep the ternary expression on a single line.

romoore replied:
Done

ianegordon commented on components/TextFields/src/MDCTextField.m:
```diff
@@ -378,7 +379,7 @@ - (CGSize)intrinsicContentSize {
   CGFloat underlineLabelsHeight =
       MAX(MDCCeil(self.leadingUnderlineLabel.font.lineHeight * 2.f) / 2.f,
-          MDCCeil(self.trailingUnderlineLabel.font.lineHeight * 2.f) / 2.f);
+              MDCCeil(self.trailingUnderlineLabel.font.lineHeight * 2.f) / 2.f);
```
Not sure about this indenting.  +4 spaces after continuing a line.  This looks like +8.

romoore replied:
Done

ianegordon commented on components/ButtonBar/src/private/MDCAppBarButtonBarBuilder.m:
```diff
@@ -82,7 +82,7 @@ - (UIView *)buttonBar:(MDCButtonBar *)buttonBar
 
   // Take the real custom view if it exists instead of sandbag view.
-  UIView *customView = buttonItem.mdc_customView ?: buttonItem.customView;
+  UIView *customView = buttonItem.mdc_customView ? buttonItem.mdc_customView : buttonItem.customView;
```
101 characters :/
Linebreak after "customView ="

romoore replied:
Done

ianegordon commented on components/Slider/src/MDCSlider.m:
```diff
@@ -96,7 +96,7 @@ - (void)commonMDCSliderInit {
 
 - (void)setTrackBackgroundColor:(UIColor *)trackBackgroundColor {
-  _thumbTrack.trackOffColor = trackBackgroundColor ?: [[self class] defaultTrackOffColor];
+  _thumbTrack.trackOffColor = trackBackgroundColor ? trackBackgroundColor : [[self class] defaultTrackOffColor];
```
Linebreak after ".trackOffColor =".

romoore replied:
Done

ianegordon commented on components/TextFields/src/MDCTextInputControllerDefault.m:
```diff
@@ -493,7 +493,8 @@ - (void)movePlaceholderToUp:(BOOL)isToUp {
       self.textInput.placeholderLabel.textColor =
-          self.previousPlaceholderColor ?: self.inlinePlaceholderColor;
+          self.previousPlaceholderColor
+            ? self.previousPlaceholderColor : self.inlinePlaceholderColor;
```
I'm beginning to question the utility of the ternary operator without an implicit second argument.  Once you end up on multiple lines it gets hard to parse.

Consider:
if (self.previousPlaceholderColor) {
  self.textInput.placeholderLabel.textColor = self.previousPlaceholderColor;
} else {
  self.textInput.placeholderLabel.textColor = self.inlinePlaceholderColor;
}

romoore replied:
Done

ianegordon commented on components/TextFields/src/MDCTextInputControllerDefault.m:
```diff
@@ -1044,7 +1047,7 @@ - (void)setErrorText:(NSString *)errorText
     NSString *valueString = @"";
 
-    if (self.textInput.text > 0) {
+    if (self.textInput.text) {
```
Was the original intention (self.textInput.text.length > 0) ?

willlarche replied:
Typo. Should be .length

ianegordon reviewed (changes requested):
I find ternary operators split across lines very hard to parse.  Can we discuss this offline?
